micro_sam/sam_3d_wrapper.py

```diff
@@ -0,0 +1,151 @@
+            x = blk(x, d_size)
+
+        x = self.image_encoder.neck(x.permute(0, 3, 1, 2))
+
+        return x
+
+
+class NDBlock(nn.Module):
+    def __init__(
+            self,
+            dim: int,
+            num_heads: int,
+            Block: nn.Module,
+            norm_layer: Type[nn.Module] = nn.LayerNorm,
+            **kwargs
+    ):
+        super().__init__()
+        self.Block = Block
+        
+        self.adapter_channels = 384
+        self.adapter_linear_down = nn.Linear(dim, self.adapter_channels, bias=False)
+        self.adapter_linear_up = nn.Linear(self.adapter_channels, dim, bias=False)
+        self.adapter_conv = nn.Conv3d(self.adapter_channels, self.adapter_channels, kernel_size=(3, 1, 1), padding='same')
+        self.adapter_act = nn.GELU()
+        self.adapter_norm = norm_layer(dim)
+
+        self.adapter_linear_down_2 = nn.Linear(dim, self.adapter_channels, bias=False)
+        self.adapter_linear_up_2 = nn.Linear(self.adapter_channels, dim, bias=False)
+        self.adapter_conv_2 = nn.Conv3d(self.adapter_channels, self.adapter_channels, kernel_size=(3, 1, 1), padding='same')
+        self.adapter_act_2 = nn.GELU()
+        self.adapter_norm_2 = norm_layer(dim)
+
+    def forward(self, x: torch.Tensor, d_size) -> torch.Tensor:
+        b_size, hw_size = x.shape[0], x.shape[1]
+        
+        # 3D adapter
+        shortcut = x
+        x = self.adapter_norm(x)
+        x = self.adapter_linear_down(x)
+        x = x.contiguous().view(int(b_size/d_size), d_size, hw_size, hw_size, self.adapter_channels)
+        x = torch.permute(x, (0, -1, 1, 2, 3))
+        x = self.adapter_conv(x)
+        x = torch.permute(x, (0, 2, 3, 4, 1))
+        x = x.contiguous().view(b_size, hw_size, hw_size, self.adapter_channels)
+        x = self.adapter_act(x)
+        x = self.adapter_linear_up(x)
+        x = shortcut + x
+        # end 3D adapter
+        
+        shortcut = x
+        x = self.norm1(x)
+        # Window partition
+        if self.window_size > 0:
+            H, W = x.shape[1], x.shape[2]
+            x, pad_hw = window_partition(x, self.window_size)
+
+        x = self.attn(x)
+        # Reverse window partition
+        if self.window_size > 0:
+            x = window_unpartition(x, self.window_size, pad_hw, (H, W))
+
+        x = shortcut + x
+        
+        # 3D adapter
+        shortcut = x
+        x = self.adapter_norm_2(x)
+        x = self.adapter_linear_down_2(x)
+        x = x.contiguous().view(int(b_size/d_size), d_size, hw_size, hw_size, self.Block.adapter_channels)
+        x = torch.permute(x, (0, -1, 1, 2, 3))
+        x = self.adapter_conv_2(x)
+        x = torch.permute(x, (0, 2, 3, 4, 1))
+        x = x.contiguous().view(b_size, hw_size, hw_size, self.Block.adapter_channels)
+        x = self.adapter_act_2(x)
+        x = self.adapter_linear_up_2(x)
+        x = shortcut + x
+        # end 3D adapter
+        
+        x = x + self.mlp(self.norm2(x))
```

Comment:
Some of this is also in `self.Block` / `self.block`.

Reply:
found some other inconsistencies regarding this